### ISSUE TITLE
Proposed RFC Feature simplify EmotionFX mesh

### ISSUE THREAD

**Issue by pollend** [o3de/sig-simulation#18]
# simplify EmotionFX mesh

### Summary

The current EmotionFX mesh has a lot of functionality with the intention of being extendable but introduces a lot of complexity in how its currently used in the engine. 

### What is the relevance of this feature?

EmotionFX maintains a mesh objects with a copy of the of the data from RPI Mesh. A lot of this data is mainly used in the editor and is not applicable to the final output of the engine. There is a separate AtomActor and SkinnedMesh currently in atom that handles this interaction. 

As it stands for the short term it makes a lot of sense to simplify the interface and drop a lot of the existing functionality in the mesh object that is not directly related to the editor. There is a lot of functional code still in mesh that is not easily removed at the moment but simplifying the interface seems like a good start.

Any modifications to the mesh are not directly applicable to Actor since the Position, Normals, Tangents, UVCoords, Bitanges are discard when constructing the ActorAsset. These attributes at the moment are mainly used for displaying debug information about the mesh(normals of the faces, tangents, etc ... )The Original vertex indexes are used to constructed the weights for the skinned mesh. some of the current modifications supported by the current mesh would cause this data to be out of sync with the current AtomActor.

The current mesh supports:
- multiple copies of attributes so its possible to have multiple attributes of the same type. you can query an attribute by its occurrence but only 1 attribute is every defined in the current implement.
- swap and shuffle vertices in the mesh. i.e shifting the position of faces/color in the mesh
- removing vertices in the mesh 

### Feature design description

- only 1 attributes of any one type is supported no need to support a potential case where multiple attributes of the same type are present. i.e Position, Tangent, Position, Tangent.
- use traits to describe the type information for the given attribute. 
    - Position Vector3
    - VertexNumbers u32
    - Normal Vector3
    - UVCoords Vector2
    - Tangets Vector4
    - Bitangent Vector3
- data from the atom mesh is copied into the attribute and only allow for modifying the fixed values in place
- there is a restricted set of attributes supported by Emotion Mesh, don't expect this to be extendable in any way.  

### What are the advantages of the feature?

- Explain the advantages for someone to use this feature

Simplifies the overall interface and shrinks the scope of the current mesh implementation. limits what can be used with the mesh. should make it more easier to understand how the underlying code is used since the current implementation is very hard to follow. 

- Explain any disadvantages for someone to use this feature

This is more of a short term patch since the current implementation is in contention with atom. I think some of this communication can be accomplished with the EBus system. a lot of this CPU driven logic is replicated on the atom side and done on the GPU. A lot of this logic shouldn't need to be replicated like this. Ideally in the long term this should be removed.

relevant PR: https://github.com/o3de/o3de/pull/10603

**Comment by amzn-jillich** [comment]
Thanks for writing up this RFC @pollend.

One problem that I see with the proposal is that afterwards only 1x attribute layer per type is supported. This is ok for positions e.g. but not for UV/tangent/bitangent layers as there might be multiple of them.

The only 2x reasons why the EMFX mesh class is still present are:
1. Server usage: Some game servers don't have a GPU (or with limited computing power) and need to deform (skinning, morphing) the meshes on CPU for detecting hits e.g.. There are plans by the Atom team to implement a mesh that is optimized for main memory and can be deformed with the CPU, but this is not for the short-term.
2. The second reason is debug rendering in the Animation Editor/main editor: Currently wireframe, normals, tangents rendering etc. is using software skinning and morphing and the AUX geom helper renderer. We will need to port all of these things as well before we can actually remove the EMFX mesh.

As the EMFX mesh class is planned to be deprecated generally in the mid- to long-term, and the old code has been battle tested on all platforms for nearly two decades, I am not sure if the risk of introducing potential new issues and breaking existing customers is worth right now - it would be great if these ideas and efforts would directly flow into the new CPU optimized Atom mesh though in my eyes.

**Comment by pollend** [comment]
> Thanks for writing up this RFC @pollend.
> 
> One problem that I see with the proposal is that afterwards only 1x attribute layer per type is supported. This is ok for positions e.g. but not for UV/tangent/bitangent layers as there might be multiple of them.
> 
> The only 2x reasons why the EMFX mesh class is still present are:
> 
>     1. Server usage: Some game servers don't have a GPU (or with limited computing power) and need to deform (skinning, morphing) the meshes on CPU for detecting hits e.g.. There are plans by the Atom team to implement a mesh that is optimized for main memory and can be deformed with the CPU, but this is not for the short-term.
> 
>     2. The second reason is debug rendering in the Animation Editor/main editor: Currently wireframe, normals, tangents rendering etc. is using software skinning and morphing and the AUX geom helper renderer. We will need to port all of these things as well before we can actually remove the EMFX mesh.
> 
> 
> As the EMFX mesh class is planned to be deprecated generally in the mid- to long-term, and the old code has been battle tested on all platforms for nearly two decades, I am not sure if the risk of introducing potential new issues and breaking existing customers is worth right now - it would be great if these ideas and efforts would directly flow into the new CPU optimized Atom mesh though in my eyes.

yea a lot of that code is working if hard to use. wondering if it would be good to just remove a lot of those methods that are unused in the current class.

**Comment by amzn-rhhong** [comment]
I think @amzn-jillich sum it up pretty well. I believe the long term goal is to get rid of maintain a mesh class for EMFX, but those two specific use cases blocking us from doing so. Also the ability to do software skinning and perform ray cast for hit detection depends on the emfx::mesh as well.

I agree emfx::mesh class for the short term should be simplified. https://github.com/o3de/o3de/pull/10603 this draft changes looks fine with me, my only concern is whether those function that's removed has no usage and no usage in the future. Would like to hear from @amzn-tommy or any atom folks to check it too.

**Comment by amzn-tommy** [comment]
I second what Ben suggested. Removing dead/unused code is great, but I don't think it's worth making other improvements to the existing mesh class when the existing code is battle tested, unless they have a noticeable impact on the end user like a performance improvement.

**Comment by lgleim** [comment]
This was discussed in today's Issue Triage meeting. @amzn-jillich will follow up with the Graphics teams and update here.

**Comment by amzn-tommy** [comment]
CPU friendly access to vertex data from Atom models is not scheduled yet, but it's going to be relatively near term in priority after some mesh feature processor refactoring and hardware instancing. Currently, we assume that positions exist as 3 floats in the model asset, and we cast the data directly to Vector3f when accessing it on the CPU. As soon as we go to change how the data is compressed, even something as simple as switching to 16 bit floats, we will need to modify how the data is accessed on the CPU since the current method will break. So that will be a good time to address how mesh data is accessed on the CPU, to minimize the number of breaking changes.

And compressing our vertices is high priority, because with large models people run out of GPU memory pretty quickly with our current setup.

**Comment by lgleim** [comment]
This RFC was discussed in today's issue triage. @amzn-jillich will follow up with a comment.

**Comment by amzn-jillich** [comment]
Hi @pollend, first of all, thank you very much for creating this RFC and spending so much time working to make O3DE better. It's because of people like you that O3DE will become a success and we really value your input. In this particular case, after some discussion and reflection, we have come to the conclusion that the risk and the time investment outweighs the potential benefit here. As this change will break current users while the system in the mid-term is being deprecated, decreasing its functionality is not something we think is a good investment and think your time could be better spent elsewhere. We would love to see you spend your time on something with a higher impact and welcome ideas on other areas to work on. We will leave this RFC in a pending state for now and follow-up on the planned deprecation work as the Atom replacement comes online. Thanks again for your time and again apologies we aren't able to move forward with this.

**Comment by moraaar** [comment]
@hultonha please provide a follow up on this. Thanks.

**Comment by hultonha** [comment]
Tagging @amzn-tommy and @galibzon to see if there's an update on the progress of what @amzn-tommy mentioned above

> CPU friendly access to vertex data from Atom models is not scheduled yet, __but it's going to be relatively near term in priority after some mesh feature processor refactoring and hardware instancing__

We decided to hold off on moving forward with this RFC due to the upcoming work from Atom to stop any wasted or duplicated effort but it would be nice to know an estimated delivery date (or even a date for a date (DFAD)) on when that work might be starting up?

Thanks!
